Nav - Wrong direction check

pfeiferj opened the pull request:
**Problem:**
If you somehow get turned around on a nav-route you will not be redirected while you are backtracking along already driven portions of the route. An example would be you came across a downed tree and needed to turn around. In that case you would not have gone off the route enough to cause a re-route because you likely turned around on the road itself.

**Solution:**
This introduces a simple solution that just tracks if your distance remaining is increasing and then triggers a re-route. This should be a simpler implementation than calculating your current heading and comparing that against the heading of the route. However, this may take longer to trigger the re-route than checking the headings.

pfeiferj committed:
wrong direction check

rchybicki commented:
Tested this today, rerouted beautifully.

sshane commented:
@pd0wm I'm not too familiar with the routing code, wanted to see what you thought of this. I can check it out if you are too busy though!

pd0wm commented:
When I originally wrote this code I had something in mind that looks at the heading between the previous and the next point. But I guess something like this works as well!

Maybe it's cleaner to keep track of `distance_difference` in `send_instruction()` so you only need to introduce one new variable.

pfeiferj commented:
Ok, refactored it some.
* Moved the difference calculation over to send_instruction and stored it instead of the total distance.
* Went ahead and renamed the variable to distance_delta as well because distance_difference just sounds weird.

I'm not sure there's a way to get down to a single new variable as we need to keep track of the closest we've been to the destination and we're still using the delta in a different part of the code. I also kept using the total distance as it seems easier to keep track of when we need to reset the smallest distance value, i.e. once upon receiving the route rather than at every instruction.

pd0wm commented on selfdrive/navd/navd.py:
```diff
@@ -273,6 +276,11 @@ def send_instruction(self):
     msg.navInstruction.timeRemainingTypical = total_time_typical
 
+    # Compute if we are going the wrong way
+    if self.smallest_distance_remaining is None or total_distance < self.smallest_distance_remaining:
```
If you use `math.inf` instead of None you can replace this by a simple ` self.smallest_distance_remaining = min( self.smallest_distance_remaining, total_distance)`

pd0wm commented:
Since `total_distance` is constant during the whole route it still feels like you should be able to rewrite the math so you only need to store a single variable. However, if it becomes hard to follow it might be better to store two.

pd0wm commented:
Probably also nice to add a reroute reason to the `cloudlog.warning` in `compute_route()` then the comma team can track this in their backend and see if there are any issues with the new logic. If I didn't miss anything, valid reasons can be `new destination`, `wrong direction` or `distance to route`

pfeiferj commented:
@pd0wm updated with your suggestions. To add a reason to the warning I ended up doing some re-organizing of the recompute conditions:
* Pulled the wrong direction check into its own function.
* Pulled the current should_recompute distance check logic into a new function.
* Pulled the no recompute while in a tunnel logic into should_recompute.
* Pulled the new destination check into should_recompute
* should_recompute calls the distance and wrong direction checks
* should_recompute now returns a reason

I also made some of the variables a little more verbose so it's more clear that they are _remaining_ values

sshane commented on selfdrive/navd/navd.py:
```diff
@@ -327,7 +338,33 @@ def reset_recompute_limits(self):
+    if not self.gps_ok and self.step_idx is not None:
+      return False, None
+
+    if new_destination != self.nav_destination:
+      cloudlog.warning(f"Got new destination from NavDestination param {new_destination}")
+      return True, RECOMPUTE_REASON.NEW_DESTINATION
+
+    if self.too_far_from_route():
+      return True, RECOMPUTE_REASON.DISTANCE_TO_ROUTE
+
+    if self.going_wrong_direction():
+      return True, RECOMPUTE_REASON.WRONG_DIRECTION
+
+    return False, None
```
Is this required for this PR?

pfeiferj replied:
not required at all, was just for the suggestion of logging the reason for recompute. I'll work on pulling it out

pfeiferj commented:
closing, if i get back around to this i'll just open a new PR
